### PR TITLE
substitutions keys must be in varkeys

### LINKED ISSUE #848: Check substitutions dict for phantom varkeys
It's easy to forget to update the substitutions dict when changing a variable's varkey, or even to just type it in wrong. It might therefore be nice to have some kind of check to ensure that every entry in the substitutions dict is a varkey in the model that it's being used for. Sorry, I know there are concerns about model building speed and this won't help but it shouldn't be a particularly hard fix.

e.g.
```python
rho = Variable('\\rho')
...

substitutions = {
    ...
    'rho': 1.225,
    ...
}

m = Model(objective, constraints, substitutions)
# Should return an error, something like "The varkey 'rho' does not appear in model"
```
This will at least prompt a user to check the varkey for rho and notice it is wrong.

### PR DESCRIPTION
**Pull request by bqpd**
closes #848

### CODE REVIEW

**Comment by bqpd** [comment]
@pgkirsch, see if this meets your needs!

**Comment by whoburg** [comment]
@bqpd, this seems to be getting commits from another PR, should it be rebased?

**Comment by bqpd** [comment]
Naw, squashing will fix all that. (but yeah I probably shoulda rebased)

**Comment by pgkirsch** [comment]
Works perfectly, thanks!

**Comment by bqpd** [comment]
@whoburg, ready for review

File: gpkit/constraints/set.py [hunk 1]
```diff
@@ -55,8 +55,19 @@ def __init__(self, constraints, substitutions=None):
                 if hasattr(self[i], "substitutions"):
                     self.substitutions.update(self[i].substitutions)
         self.reset_varkeys()
-        self.substitutions.update({self[k].key: v for k, v in subs.items()
-                                   if k in self.varkeys})
```
Comment (whoburg) [inline gpkit/constraints/set.py]: let's keep the deleted lines above, commented out, with a note about timing? (I didn't realize the alternative implementation would be so many more lines. Totally open to moving back to @bqpd's original short and sweet version as long as it's not slow.)

File: gpkit/keydict.py [hunk 2]
```diff
@@ -48,7 +49,23 @@ def update(self, *args, **kwargs):
...
+                    if key.veckey:
+                        key = key.veckey
+                    elif len(keys) > 1:
+                        raise ValueError("substitution key '%s' was ambiguous;"
+                                         " .variables_byname('%s') will show"
+                                         " which variables it may refer to."
+                                         % (key, key))
+                else:
+                    raise KeyError("key '%s' does not refer to any varkey in"
+                                   " this ConstraintSet" % key)
```
Comment (whoburg) [inline gpkit/keydict.py]: what an appropriate use of `KeyError`

File: gpkit/constraints/set.py (hunk 1 above)
Comment (bqpd) [inline gpkit/constraints/set.py]: It goes back to being short and sweet in #873 :)

**Comment by whoburg** [comment]
ok, I reviewed commit cac69dce0495145b07521ed2a0f50265f77ccaf3 and it looks good. Now realizing I probably reviewed out of order; this is not ready to merge because it includes the commits from #858, which depends on #854.

Getting there...

**Comment by bqpd** [comment]
Naw, this doesn't include #858, and I already merged #854...

**Comment by whoburg** [comment]
The commits tab (which has 6 commits) is pretty confusing then...

**Pull request merged**